### PR TITLE
Add experimental alternative LLPC pipeline cache support

### PR DESCRIPTION
**Pull request by kuhar**
This is off by default and can be enabled by setting `LLPC_USE_EXPERIMENTAL_SHADER_CACHE_PIPELINES=1`.

The PR with the LLPC changes: https://github.com/GPUOpen-Drivers/llpc/pull/556.

### CODE REVIEW

**Comment by amdvlk-admin** [comment]
Can one of the admins verify this patch?

**Comment by kuhar** [comment]
Fixed build options' naming per @jinjianrong's sugestions

**Comment by JaxLinAMD** [comment]
ok to test

**Comment by JaxLinAMD** [comment]
retest this please

**Comment by JaxLinAMD** [comment]
test this please

**Comment by JaxLinAMD** [comment]
retest this please

**Comment by JaxLinAMD** [comment]
retest this please

File: icd/api/pipeline_compiler.cpp [hunk 1]
```diff
@@ -697,6 +790,53 @@ VkResult PipelineCompiler::CreateGraphicsPipelineBinary(
...
+        {
+            pCreateInfo->elfWasCached = true;
+            shouldCompile             = false;
+            m_cacheHits++;
+            pCreateInfo->pipelineFeedback.hitApplicationCache = true;
+        }
+    }
+#endif // XGL_USE_EXPERIMENTAL_SHADER_CACHE_PIPELINES
+
+#if VKI_USE_PAL_PIPELINE_CACHE
```
Comment (amingriyue) [inline icd/api/pipeline_compiler.cpp:839]: Can we make mutex between XGL_USE_EXPERIMENTAL_SHADER_CACHE_PIPELINES and VKI_USE_PAL_PIPELINE_CACHE?
Here should be \#else ?

Comment (kuhar) [reply icd/api/pipeline_compiler.cpp:839]: These two are mutually exclusive, i.e., VKI_USE_PAL_PIPELINE_CACHE is defined to true iff XGL_USE_EXPERIMENTAL_SHADER_CACHE_PIPELINES is set.

Do you suggest to remove the VKI_USE_PAL_PIPELINE_CACHE and have if else everywhere?

Comment (amingriyue) [reply icd/api/pipeline_compiler.cpp:839]: Yes, it's my opinion.
We don't need introduce VKI_USE_PAL_PIPELINE_CACHE more， and tow cache are enabled at same time, the performance could be impacted.

Comment (kuhar) [reply icd/api/pipeline_compiler.cpp:839]: I'd rather keep this macro to stay minimize the amount of changes compared to the proprietary driver. VKI_USE_PAL_PIPELINE was introduced introduced in these sources 2 years ago by a merge by Ping Fu.

Let me know if I should remove it anyway or can keep it.

Comment (amingriyue) [reply icd/api/pipeline_compiler.cpp:839]: VKI_USE_PAL_PIPELINE should already be removed when we open source pipeline binary cache in my memory.
I also grep it in whole amdvlk repo, I didn't see it.
Anything I still missed?

Comment (kuhar) [reply icd/api/pipeline_compiler.cpp:839]: Hmm, perhaps our sources diverged here. If that's the case, I'll remove it in favor of #if #else, like you recommended.

**Comment by kuhar** [comment]
Ping.

File: icd/api/pipeline_compiler.cpp [hunk 2]
```diff
@@ -763,23 +897,28 @@ VkResult PipelineCompiler::CreateGraphicsPipelineBinary(
...
 
-    if ((m_pBinaryCache != nullptr) &&
-        (isInternalCacheHit == false) &&
+#else // XGL_USE_EXPERIMENTAL_SHADER_CACHE_PIPELINES
+
+    if ((pPipelineBinaryCache != nullptr) &&
+        (isUserCacheHit == false) &&
         (result == VK_SUCCESS))
     {
-        cacheResult = m_pBinaryCache->StorePipelineBinary(
```
Comment (JaxLinAMD) [inline icd/api/pipeline_compiler.cpp:782]: if ((m_pBinaryCache != nullptr) &&
        (isInternalCacheHit == false) &&
        (result == VK_SUCCESS))
    {
        cacheResult = m_pBinaryCache->StorePipelineBinary(
            pCacheId,
            *pPipelineBinarySize,
            *ppPipelineBinary);

        VK_ASSERT(Util::IsErrorResult(cacheResult) == false);
}
this piece code exists in original code path, it's gone, do you delete it intentionally? without the piece of code, out internal tests are failed.

Comment (kuhar) [reply icd/api/pipeline_compiler.cpp:782]: This was an unintentional change made when porting the code. Thank for catching it!

Comment (kuhar) [reply icd/api/pipeline_compiler.cpp]: Also removed unintended changes in the if condition.

**Comment by kuhar** [comment]
Closing after the most recent discussion.

**Pull request closed without merging**